tools/hive-bigquery/README.md

```diff
@@ -0,0 +1,65 @@
+
+# Before you begin
+1. Install gcloud SDK on the source cluster by following these [instructions](https://cloud.google.com/sdk/install).
+2. [Create a service account](https://cloud.google.com/iam/docs/creating-managing-service-accounts#creating_a_service_account) and grant the following roles to the service account by following these [instructions](https://cloud.google.com/iam/docs/granting-roles-to-service-accounts#granting_access_to_a_service_account_for_a_resource)
+	1. Google Cloud Storage - storage.admin
+	2. BigQuery - bigquery.dataEditor and bigquery.jobUser
+	3. Cloud SQL - cloudsql.admin
+3. [Download](https://cloud.google.com/iam/docs/creating-managing-service-account-keys#creating_service_account_keys) the service account key file and set up the environment variable by following these [instructions](https://cloud.google.com/docs/authentication/getting-started#setting_the_environment_variable). 
+
+	```
+	export GOOGLE_APPLICATION_CREDENTIALS="[PATH_TO_KEY_FILE]"
+	```
+Note: Saving credentials in environment variables is convenient, but not secure - consider a more secure solution such as [Cloud KMS](https://cloud.google.com/kms/) to help keep secrets safe.
+
+# Usage
+
+1. Clone the repo
+```
+git clone https://github.com/pkattamuri/professional-services.git
+cd professional-services/
+git checkout feature/hive-bigquery
+cd tools/hive-bigquery/
+```
+2. Install prerequisites such as python, pip, virtualenv and Cloud SQL proxy
+```
+sh prerequisites/prerequisites.sh
+```
+3. Activate the virtual environment
+```
+virtualenv env
+source env/bin/activate
+pip install -r prerequisites/requirements.txt
+```
+4. [optional] Use an existing Cloud SQL instance for tracking the progress of migration or launch a new one. Run the below command to create an instance. This will output the connection name of the instance which is of use in the next steps.
+```
+sh prerequisites/create_sql_instance.sh <INSTANCE_NAME> <DATABASE_NAME>
+```
+5. Start the Cloud SQL proxy and establish a connection to the Cloud SQL instance
+```
+./cloud_sql_proxy -instances=<INSTANCE_CONNECTION_NAME>=tcp:<PORT>
+```
+6. Run hive_to_bigquery.py
+ ```
+ # Refer init_script.py for arguments descriptions
+ python hive_to_bigquery.py \
+ --hive-server-host <HIVE_SERVER_HOSTNAME> \
+ --hive-server-port <HIVE_SERVER_PORT> \
+ --hive-server-username <HIVE_SERVER_USER> \
+ --hive-database <HIVE_DB_NAME> \
+ --hive-table <HIVE_TABLE_NAME> \
+ --project-id <GCP_PROJECT_ID> \
+ --bq-dataset-id <BQ_DATASET_ID> \
+ --bq-table <BQ_TABLE_NAME> \
+ --bq-table-write-mode [create|append|overwrite] \
+ --gcs-bucket-name <GCS_BUCKET_NAME> \
+ --incremental-col <INCREMENTAL_COLUMN_NAME> \
+ --use-clustering [False|True] \
+ --tracking-database-host <TRACKING_DATABASE_HOST> \
+ --tracking-database-port <TRACKING_DATABASE_PORT> \
+ --tracking-database-user <TRACKING_DATABASE_USER> \
+ --tracking-database-db-name <TRACKING_DB_NAME>
+```
```

Comment:
Following these instructions results in a critical failure with a timeout when trying to connect to Cloud SQL instance. This is probably due to an issue with the cloud sql proxy setup. It is probably worthwhile to explain how to get this up and running after encountering this failure in a "Troubleshooting" section of these docs.

Reply:
I checked it once again by running the code. Did you face any particular issue or you need me to add troubleshoot section for Cloud SQL in general?

Reply:
I am getting the same error that jaketf has reported above. Could you please provide more information on how to test the cloud sql connectivity through the cloud sql proxy?

Reply:
@shsunil11 I tried both localhost and 127.0.0.1 for tracking-database-host-parameter. It is working fine. What parameter did you provide initially when you were getting the error? I can try and debug that once.

Reply:
Probably don't need to debug, I was passing the connection name of the MySQL tracking database. Maybe in init_script.py and in the documentation, it would be good to explain what values should be passed for the TRACKING-DATABASE-HOST and TRACKING-DATABASE-PORT.

Reply:
Just wanted to check on the following points.

1.
A step should be added to install the cloud_sql_proxy, as pointed out by jaketf

2. 
Should the cloud_sql_proxy be executed in background with nohup?

3.
I had to run the prerequisites.sh as sudo, otherwise it would not work

4.
Looks like the create_sql_instance.sh accepts a third argument for REGION. The instruction does not mention this argument.

5.
Would be good to add the command to connect to the mysql tracking instance from the command line.
Something like (from the instance from which cloud_sql_proxy was launched)
mysql -h localhost -P <port> -u root -p

Where port is the port on which the cloud_sql_proxy was launched

6.
Is there a plan to finalize and integrate this branch with the master branch in the near future?

Reply:
Facing issue with squashing the below two commits 3265d6a and 3299481 .They accidentally got a different user name and hence failed the cla. I think you are not able to view the changes I made since then like point no 2 and 4
1. prerequisites.sh script installs cloud_sql_proxy.
2. trailing & has been added to the cloud_sql_proxy execution command.
3. I'll update it.
4. Already updated in the README file
5. Will add this command to test the connection.
6. Yes.

Reply:
@shsunil11 @jaketf  i've made the additional changes in the README and init_script.py regarding the host and port of cloud-sql-proxy and how to test the connection using mysql client. Please check once and if everything is fine, I can mark this as resolved.

Reply:
Hi pkattamuri,
Few points 
1. When loading a hive table with a long name, we ran into a MySql Issue. Looks like MySql has a table name length limit of 64 chars. So if the db_name+table_name+plus_some_suffix exceeds 64 chars, the script fails.
Not sure what is the best solution, but I was wondering, do you need to create a separate tracking table for each hive table, can this be stored in a single tracking table, the tracking table has a column for table_name anyway. Just a thought.

2. When I tried to load the same hive table for a second time into a different BQ table, i got the error that the tracking table already exists. 

3.  The few updates to the README (6 points listed above). Have they been added to the README?

Reply:
Thanks @shsunil11 
1. I will look into this and get it resolved.
2. Are you using "append" for the --bq-table-write-mode argument when you are loading again? Can you provide more information on what parameters you have passed.
3. The cla issue is not resolved yet, so you might not be able to view the changes I have done in the README file. I will let you know on this.